File: sdk/reproit_flutter/test/fingerprint_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:reproit_flutter/reproit_flutter.dart';

// Unit tests for the PII-safe input fingerprint (tier-3 on-error context).
// Parity cases mirror the other four SDKs: capture FEATURES, never raw values.
void main() {
  group('fingerprintValue, PII-safe features (not values)', () {
    test('"José🎉" counts code points and is unicode + emoji', () {
      final r = ReproIt.fingerprintValue('José🎉');
      expect(r['len'], 5); // code points, not UTF-16 length
      expect(r['charset'], 'unicode');
      expect(r['hasEmoji'], true);
      expect(r['isEmpty'], false);
      expect(r['isRtl'], false);
    });

    test('"12345" is numeric', () {
      final r = ReproIt.fingerprintValue('12345');
      expect(r['charset'], 'numeric');
      expect(r['len'], 5);
      expect(r['hasEmoji'], false);
    });

    test('"hello" is ascii', () {
      expect(ReproIt.fingerprintValue('hello')['charset'], 'ascii');
    });

    test('"" isEmpty', () {
      final r = ReproIt.fingerprintValue('');
      expect(r['isEmpty'], true);
      expect(r['len'], 0);
      expect(r['charset'], 'ascii');
    });

    test('whitespace-only isEmpty', () {
      expect(ReproIt.fingerprintValue('   ')['isEmpty'], true);
    });

    test('Arabic string isRtl + unicode', () {
      final r = ReproIt.fingerprintValue('مرحبا');
      expect(r['isRtl'], true);
      expect(r['charset'], 'unicode');
      expect(r['hasEmoji'], false);
    });

    test('Hebrew string isRtl', () {
      expect(ReproIt.fingerprintValue('שלום')['isRtl'], true);
    });

    test('Turkish dotless i is unicode not ascii', () {
      final r = ReproIt.fingerprintValue('ıstanbul');
      expect(r['charset'], 'unicode');
      expect(r['isRtl'], false);
    });

    test('312-char name reports exact length', () {
      final r = ReproIt.fingerprintValue('a' * 312);
      expect(r['len'], 312);
      expect(r['charset'], 'ascii');
    });

    test('never echoes the raw value', () {
      const raw = 'secret-pii-value';
      expect(jsonEncode(ReproIt.fingerprintValue(raw)).contains(raw), isFalse);
    });
  });

  // ---- v2 features (bytes / scripts / combining / zero-width / newline / ws) --
  // Parity with the web reference (reproit-web.js) and its fingerprint_test.js.
  group('fingerprintValue v2 features', () {
    test('bytes is UTF-8 length, distinct from code-point len', () {
      // J o s é(2B) 🎉(4B) -> 9 bytes, 5 code points.
      final r = ReproIt.fingerprintValue('Jos\u{00E9}\u{1F389}');
      expect(r['len'], 5);
      expect(r['bytes'], 9);
      expect(
          ReproIt.fingerprintValue('hello')['bytes'], 5); // ascii: bytes == len
    });

    test('graphemes counts user-visible clusters', () {
      expect(ReproIt.fingerprintValue('hello')['graphemes'], 5);
      expect(ReproIt.fingerprintValue('e\u{0301}')['len'], 2);
      expect(ReproIt.fingerprintValue('e\u{0301}')['graphemes'], 1);
      expect(ReproIt.fingerprintValue('👨‍👩‍👧‍👦')['graphemes'], 1);
    });

    test('scripts lists buckets present, sorted, mixed-script', () {
      expect(ReproIt.fingerprintValue('hello')['scripts'], ['Latin']);
      expect(ReproIt.fingerprintValue('hello')['bytes'], 5);
      // Arabic "مرحبا".
      final ar =
          ReproIt.fingerprintValue('\u{0645}\u{0631}\u{062D}\u{0628}\u{0627}');
      expect(ar['scripts'], ['Arabic']);
      expect(ar['isRtl'], true);
      // Mixed Latin + Arabic.
      expect(
        ReproIt.fingerprintValue(
            'hi \u{0645}\u{0631}\u{062D}\u{0628}\u{0627}')['scripts'],
        ['Arabic', 'Latin'],
      );
      // CJK "日本語".
      expect(
        ReproIt.fingerprintValue('\u{65E5}\u{672C}\u{8A9E}')['scripts'],
        ['CJK'],
      );
      // Digits are no script.
      expect(ReproIt.fingerprintValue('12345')['scripts'], <String>[]);
    });

    test('hasNewline detects LF and CR', () {
      expect(ReproIt.fingerprintValue('line1\nline2')['hasNewline'], true);
      expect(ReproIt.fingerprintValue('a\rb')['hasNewline'], true);
      expect(ReproIt.fingerprintValue('oneline')['hasNewline'], false);
    });

    test('hasZeroWidth detects invisible code points', () {
      expect(
          ReproIt.fingerprintValue('a\u{200B}b')['hasZeroWidth'], true); // ZWSP
      expect(ReproIt.fingerprintValue('ab')['hasZeroWidth'], false);
    });

    test('hasCombiningMarks detects decomposed accents', () {
      // e + combining acute.
      expect(ReproIt.fingerprintValue('e\u{0301}')['hasCombiningMarks'], true);
      expect(ReproIt.fingerprintValue('e')['hasCombiningMarks'], false);
      // Precomposed é.
      expect(ReproIt.fingerprintValue('\u{00E9}')['hasCombiningMarks'], false);
    });

    test('leadingTrailingWhitespace flags edge whitespace', () {
      expect(ReproIt.fingerprintValue(' hello')['leadingTrailingWhitespace'],
          true);
      expect(ReproIt.fingerprintValue('hello ')['leadingTrailingWhitespace'],
          true);
      expect(ReproIt.fingerprintValue('hello')['leadingTrailingWhitespace'],
          false);
      // Interior tab only.
      expect(
          ReproIt.fingerprintValue('a\tb')['leadingTrailingWhitespace'], false);
    });

    test('fpVersion is 2', () {
      expect(ReproItFingerprint.fpVersion, 2);
    });
  });
}

File: sdk/reproit-react-native/test/fingerprint.test.ts
/**
 * Unit tests for the PII-safe input fingerprint (tier-3 on-error context).
 * Parity cases mirror the other four SDKs: capture FEATURES, never raw values.
 */
import { fingerprintValue, fingerprintFields, FP_VERSION } from '../src/fingerprint';

describe('fingerprintValue, PII-safe features (not values)', () => {
  test('"José🎉" counts code points and is unicode + emoji', () => {
    const r = fingerprintValue('José🎉');
    expect(r.len).toBe(5); // code points, not UTF-16 length
    expect(r.charset).toBe('unicode');
    expect(r.hasEmoji).toBe(true);
    expect(r.isEmpty).toBe(false);
    expect(r.isRtl).toBe(false);
  });

  test('"12345" is numeric', () => {
    const r = fingerprintValue('12345');
    expect(r.charset).toBe('numeric');
    expect(r.len).toBe(5);
    expect(r.hasEmoji).toBe(false);
  });

  test('"hello" is ascii', () => {
    expect(fingerprintValue('hello').charset).toBe('ascii');
  });

  test('"" isEmpty', () => {
    const r = fingerprintValue('');
    expect(r.isEmpty).toBe(true);
    expect(r.len).toBe(0);
    expect(r.charset).toBe('ascii');
  });

  test('whitespace-only isEmpty', () => {
    expect(fingerprintValue('   ').isEmpty).toBe(true);
  });

  test('Arabic string isRtl + unicode', () => {
    const r = fingerprintValue('مرحبا');
    expect(r.isRtl).toBe(true);
    expect(r.charset).toBe('unicode');
    expect(r.hasEmoji).toBe(false);
  });

  test('Hebrew string isRtl', () => {
    expect(fingerprintValue('שלום').isRtl).toBe(true);
  });

  test('Turkish dotless i is unicode not ascii', () => {
    const r = fingerprintValue('ıstanbul');
    expect(r.charset).toBe('unicode');
    expect(r.isRtl).toBe(false);
  });

  test('312-char name reports exact length', () => {
    const r = fingerprintValue('a'.repeat(312));
    expect(r.len).toBe(312);
    expect(r.charset).toBe('ascii');
  });

  test('null / undefined treated as empty', () => {
    expect(fingerprintValue(null).isEmpty).toBe(true);
    expect(fingerprintValue(undefined).isEmpty).toBe(true);
  });

  test('never echoes the raw value', () => {
    const raw = 'secret-pii-value';
    expect(JSON.stringify(fingerprintValue(raw))).not.toContain(raw);
  });
});

describe('fingerprintValue v2 features (bytes / scripts / combining / zero-width / newline / ws)', () => {
  test('bytes is UTF-8 length, distinct from code-point len', () => {
    const r = fingerprintValue('José\u{1F389}'); // J o s é(2B) 🎉(4B) -> 9 bytes, 5 code points
    expect(r.len).toBe(5);
    expect(r.bytes).toBe(9);
    expect(fingerprintValue('hello').bytes).toBe(5); // ascii: bytes == len
  });

  test('graphemes counts user-visible clusters', () => {
    expect(fingerprintValue('hello').graphemes).toBe(5);
    expect(fingerprintValue('e\u{0301}').len).toBe(2);
    expect(fingerprintValue('e\u{0301}').graphemes).toBe(1);
    expect(fingerprintValue('👨‍👩‍👧‍👦').graphemes).toBe(1);
  });

  test('scripts lists buckets present, sorted, mixed-script', () => {
    expect(fingerprintValue('hello').scripts).toEqual(['Latin']);
    expect(fingerprintValue('hello').bytes).toBe(5);
    expect(fingerprintValue('مرحبا').scripts).toEqual(['Arabic']);
    expect(fingerprintValue('مرحبا').isRtl).toBe(true);
    expect(fingerprintValue('hi مرحبا').scripts).toEqual(['Arabic', 'Latin']);
    expect(fingerprintValue('日本語').scripts).toEqual(['CJK']);
    expect(fingerprintValue('12345').scripts).toEqual([]); // digits are no script
  });

  test('hasNewline detects LF and CR', () => {
    expect(fingerprintValue('line1\nline2').hasNewline).toBe(true);
    expect(fingerprintValue('a\rb').hasNewline).toBe(true);
    expect(fingerprintValue('oneline').hasNewline).toBe(false);
  });

  test('hasZeroWidth detects invisible code points', () => {
    expect(fingerprintValue('a\u{200B}b').hasZeroWidth).toBe(true); // ZWSP
    expect(fingerprintValue('ab').hasZeroWidth).toBe(false);
  });

  test('hasCombiningMarks detects decomposed accents', () => {
    expect(fingerprintValue('e\u{0301}').hasCombiningMarks).toBe(true); // e + combining acute
    expect(fingerprintValue('\u{00E9}').hasCombiningMarks).toBe(false); // precomposed é
    expect(fingerprintValue('e').hasCombiningMarks).toBe(false);
  });

  test('leadingTrailingWhitespace flags edge whitespace', () => {
    expect(fingerprintValue(' hello').leadingTrailingWhitespace).toBe(true);
    expect(fingerprintValue('hello ').leadingTrailingWhitespace).toBe(true);
    expect(fingerprintValue('hello').leadingTrailingWhitespace).toBe(false);
    expect(fingerprintValue('a\tb').leadingTrailingWhitespace).toBe(false); // interior only
  });

  test('FP_VERSION is exported and is 2', () => {
    expect(FP_VERSION).toBe(2);
  });
});

describe('fingerprintFields', () => {
  test('keeps the field label, drops the value, emits features', () => {
    const out = fingerprintFields([
      { field: 'email', value: 'a@b.co' },
      { field: '#1', value: '12345' },
      { field: 'note', value: '' },
    ]);
    expect(out).toHaveLength(3);
    expect(out[0].field).toBe('email');
    expect(out[1].charset).toBe('numeric');
    expect(out[2].isEmpty).toBe(true);
    // no raw values anywhere in the output
    expect(JSON.stringify(out)).not.toContain('a@b.co');
  });
});
